src/type_conversion/runpython.c

```diff
@@ -1,165 +1,132 @@
 #include "runpython.h"
-
-#include <Python.h>
-#include <emscripten.h>
-#include <node.h> // from Python
-
 #include "hiwire.h"
+#include "pyproxy.h"
 #include "python2js.h"
 
-PyObject* globals;
-PyObject* pyodide;
+#include <Python.h>
+#include <emscripten.h>
 
+static PyObject* pyodide_py;
+static PyObject* globals;
 _Py_IDENTIFIER(eval_code);
-_Py_IDENTIFIER(find_imports);
 
 JsRef
-_runPython(char* code)
+_runPythonDebug(char* code)
 {
   PyObject* py_code;
   py_code = PyUnicode_FromString(code);
   if (py_code == NULL) {
-    pythonexc2js();
-    return Js_ERROR;
+    fprintf(stderr, "runPythonDebug -- error occurred converting argument:\n");
+    PyErr_Print();
+    return Js_UNDEFINED;
   }
 
-  PyObject* ret = _PyObject_CallMethodIdObjArgs(
-    pyodide, &PyId_eval_code, py_code, globals, NULL);
+  PyObject* result = _PyObject_CallMethodIdObjArgs(
+    pyodide_py, &PyId_eval_code, py_code, globals, NULL);
 
-  if (ret == NULL) {
-    pythonexc2js();
-    return Js_ERROR;
+  if (result == NULL) {
+    fprintf(stderr, "runPythonDebug -- error occurred\n");
+    PyErr_Print();
+    return Js_UNDEFINED;
   }
-  JsRef id = python2js(ret);
-  Py_DECREF(ret);
-  return id;
-}
 
-JsRef
-_findImports(char* code)
-{
-  PyObject* py_code;
-  py_code = PyUnicode_FromString(code);
-  if (py_code == NULL) {
-    pythonexc2js();
-    return Js_ERROR;
-  }
-
-  PyObject* ret =
-    _PyObject_CallMethodIdObjArgs(pyodide, &PyId_find_imports, py_code, NULL);
+  printf("runPythonDebug -- eval_code succeeded, it returned:\n");
+  PyObject_Print(result, stdout, 0);
 
-  if (ret == NULL) {
-    pythonexc2js();
-    return Js_ERROR;
-  }
-
-  JsRef id = python2js(ret);
-  Py_DECREF(ret);
+  printf("runPythonDebug -- doing python2js(result):\n");
+  JsRef id = python2js(result);
+  Py_DECREF(result);
   return id;
 }
 
-EM_JS(int, runpython_init_js, (), {
-  Module._runPythonInternal = function(pycode)
+EM_JS(int, runpython_init_js, (JsRef pyodide_py_proxy, JsRef globals_proxy), {
+  Module.pyodide_py = Module.hiwire.get_value(pyodide_py_proxy);
+  Module.globals = Module.hiwire.get_value(globals_proxy);
+
+  // Use this to test python code separate from pyproxy.apply.
+  Module.runPythonDebug = function(code)
```

Comment:
Would it make more sense to move this to pyodide.js alongside the definition of `runPython`?

Reply:
I prefer to keep it here: the definition of `runPython` doesn't do explicit memory management, it leaves that to `PyProxy`, whereas `runPythonDebug` just does memory management / js <==> C conversion and then calls the C function.